In [ ]:
# Nome: Gabriel de Lima Miranda
# RA: 22304286

# Orientações para a Prova

Por favor, leia **atentamente** todas as instruções abaixo para realizar e enviar sua prova corretamente.

# `Deve ser entregue até as 9:40, horário que o formulário irá parar de receber respostas`
## Identificação

- **Preencha** seu nome e RA no bloco de código acima.

## Formato e Envio do Arquivo

- Envie o arquivo da prova com o formato `nome_sobrenome.ipynb` (sendo o primeiro sobrenome).
  
  **Exemplos:**
  
  - Nome: Ana Maria da Silva Sauro  
    Arquivo: `ana_maria_silva.ipynb`
    
  - Nome: Arthur Porfírio de Castro Siqueira  
    Arquivo: `arthur_porfirio.ipynb`
    
  - Nome: Marco Antônio Rodriges Fernandes  
    Arquivo: `marco_antonio_rodrigues.ipynb`
    
  - Nome: Flávia Melo Alcântara  
    Arquivo: `flavia_melo.ipynb`

- A prova deve ser enviada **exclusivamente** pelo formulário disponível [neste link](https://forms.gle/xdXZMP89KgKDS4hFA).

  <span style="color:red">**ATENÇÃO:**</span> A prova **NÃO** deve ser enviada para o e-mail do professor, apenas pelo formulário.

## Regras da Prova

- A prova pode ser realizada com consulta apenas em materiais das aulas anteriores ou anotações próprias.

- <span style="color:red">**Quem for pego colando terá a nota II.**</span>

- **Sem conversas paralelas** durante a prova, isto pode ser considerado cola.

- **Não** devem ser criados mais blocos de códigos que os disponíveis na prova (no desenvolvimento do código podem ser criados mais blocos de código, mas antes de enviar devem conter apenas os blocos de códigos que estavam disponíveis na prova).

- O notebook deve conter as saídas dos blocos de código referentes às questões que possuem gráficos e/ou valores que exibam as respostas (não é para limpar a saída dos blocos de código para me enviar).

- Cada questão vale 1 ponto
## Boa Prova!

# 1 - Importe todas as bibliotecas no bloco de código abaixo

menos 0.10 pontos da questão para cada:
- Biblioteca que for importada e não for utilizada;
- Biblioteca importada fora deste bloco de código;
- Biblioteca importada repetida.

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.decomposition import PCA
from sklearn.preprocessing import MinMaxScaler, StandardScaler
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

# 2 - Separação dos dados em X e y
A base de dados score_credito.csv tem informações de pessoas com epréstimos e no fim qual (coluna score_credito) qual o score de crédito da pessoa.

Separe os dados em X para os atributos e y para o score de crédito.

In [ ]:
dados_credito = pd.read_csv('score_credito.csv')

In [ ]:
X = dados_credito.drop('score_credito', axis=1)
y = dados_credito['score_credito']

# 3 - Dividir os dados em Treinamento (70%) e Teste (30%)

In [ ]:
X_treino, X_teste, y_treino, y_teste = train_test_split(X, y, test_size=0.3, random_state=42)

# 4 - Aplique o PCA para reduzir a dimensionalidade para 10 colunas
- Crie o X_train_pca e o X_test_pca
- Crie um scaler MinMax chamado scaler_pca
- Normalize X_train_pca (fit_transform) e X_test_pca (transform)

In [ ]:
num_componentes = 10
pca = PCA(n_components=num_componentes)

X_train_pca = pca.fit_transform(X_treino)

X_test_pca = pca.transform(X_teste)

scaler_pca = MinMaxScaler()

X_train_pca_normalized = scaler_pca.fit_transform(X_train_pca)

X_test_pca_normalized = scaler_pca.transform(X_test_pca)

# 5 - Aplique o LDA para reduzir a dimensionalidade para 3 colunas
- Crie o X_train_lda e o X_test_lda
- Crie um scaler MinMax chamado scaler_lda
- Normalize X_train_lda (fit_transform) e X_test_lda (transform)
---
- Crie um scaler chamado scaler_normal
- Normalize X_train (fit_transform) e X_test (transform)

In [ ]:
num_componentes_lda = 3
lda = LinearDiscriminantAnalysis(n_components=num_componentes_lda)

X_train_lda = lda.fit_transform(X_treino, y_treino)

X_test_lda = lda.transform(X_teste)

scaler_lda = MinMaxScaler()

X_train_lda_normalized = scaler_lda.fit_transform(X_train_lda)

X_test_lda_normalized = scaler_lda.transform(X_test_lda)

scaler_normal = MinMaxScaler()

X_train_normalized = scaler_normal.fit_transform(X_treino)

X_test_normalized = scaler_normal.transform(X_teste)

# 6 - Treine 3 modelo de kNN com os dados de Treinamento
modelos devem chamar:
- modelo_knn_normal (treinado com X_train)
- modelo_knn_pca (treinado com X_train_pca )
- modelo_knn_lda (treinado com X_train_lda)

In [ ]:
modelo_knn_normal = KNeighborsClassifier()
modelo_knn_normal.fit(X_train_normalized, y_treino)

modelo_knn_pca = KNeighborsClassifier()
modelo_knn_pca.fit(X_train_pca_normalized, y_treino)

modelo_knn_lda = KNeighborsClassifier()
modelo_knn_lda.fit(X_train_lda_normalized, y_treino)

KNeighborsClassifier()

# 7 - Treine 3 modelo de Naive Bayes com os dados de Treinamento
modelos devem chamar:
- modelo_nb_normal (treinado com X_train)
- modelo_nb_pca (treinado com X_train_pca )
- modelo_nb_lda (treinado com X_train_lda)

In [ ]:
modelo_nb_normal = GaussianNB()
modelo_nb_normal.fit(X_train_normalized, y_treino)

modelo_nb_pca = GaussianNB()
modelo_nb_pca.fit(X_train_pca_normalized, y_treino)

modelo_nb_lda = GaussianNB()
modelo_nb_lda.fit(X_train_lda_normalized, y_treino)

GaussianNB()

# 8 - Treine 3 modelo de Árvore de Decisão com os dados de Treinamento
modelos devem chamar:
- modelo_ad_normal (treinado com X_train)
- modelo_ad_pca (treinado com X_train_pca )
- modelo_ad_lda (treinado com X_train_lda)

In [ ]:
modelo_ad_normal = DecisionTreeClassifier()
modelo_ad_normal.fit(X_train_normalized, y_treino)

modelo_ad_pca = DecisionTreeClassifier()
modelo_ad_pca.fit(X_train_pca_normalized, y_treino)

modelo_ad_lda = DecisionTreeClassifier()
modelo_ad_lda.fit(X_train_lda_normalized, y_treino)

DecisionTreeClassifier()

# 9 - Treine 3 modelo de Random Forest com 10 árvores e com os dados de Treinamento
modelos devem chamar:
- modelo_rf_normal (treinado com X_train)
- modelo_rf_pca (treinado com X_train_pca )
- modelo_rf_lda (treinado com X_train_lda)


In [ ]:
modelo_rf_normal = RandomForestClassifier(n_estimators=10, random_state=42)
modelo_rf_normal.fit(X_train_normalized, y_treino)

modelo_rf_pca = RandomForestClassifier(n_estimators=10, random_state=42)
modelo_rf_pca.fit(X_train_pca_normalized, y_treino)

modelo_rf_lda = RandomForestClassifier(n_estimators=10, random_state=42)
modelo_rf_lda.fit(X_train_lda_normalized, y_treino)

RandomForestClassifier(n_estimators=10, random_state=42)

# 10 - Exiba a acurácia de todos os modelos nos dados de teste e informe qual(quais) foi(foram) o(s) melhor(es) modelo(s).

In [ ]:
def calcular_acuracia(modelo, X_test, y_test, nome_modelo):
    y_pred = modelo.predict(X_test)
    acuracia = accuracy_score(y_test, y_pred)
    print(f'Acurácia do {nome_modelo}: {acuracia:.4f}')

calcular_acuracia(modelo_knn_normal, X_test_normalized, y_teste, 'modelo_knn_normal')
calcular_acuracia(modelo_knn_pca, X_test_pca_normalized, y_teste, 'modelo_knn_pca')
calcular_acuracia(modelo_knn_lda, X_test_lda_normalized, y_teste, 'modelo_knn_lda')

calcular_acuracia(modelo_nb_normal, X_test_normalized, y_teste, 'modelo_nb_normal')
calcular_acuracia(modelo_nb_pca, X_test_pca_normalized, y_teste, 'modelo_nb_pca')
calcular_acuracia(modelo_nb_lda, X_test_lda_normalized, y_teste, 'modelo_nb_lda')

calcular_acuracia(modelo_ad_normal, X_test_normalized, y_teste, 'modelo_ad_normal')
calcular_acuracia(modelo_ad_pca, X_test_pca_normalized, y_teste, 'modelo_ad_pca')
calcular_acuracia(modelo_ad_lda, X_test_lda_normalized, y_teste, 'modelo_ad_lda')

calcular_acuracia(modelo_rf_normal, X_test_normalized, y_teste, 'modelo_rf_normal')
calcular_acuracia(modelo_rf_pca, X_test_pca_normalized, y_teste, 'modelo_rf_pca')
calcular_acuracia(modelo_rf_lda, X_test_lda_normalized, y_teste, 'modelo_rf_lda')

Acurácia do modelo_knn_normal: 0.5587
Acurácia do modelo_knn_pca: 0.7178
Acurácia do modelo_knn_lda: 0.9780
Acurácia do modelo_nb_normal: 0.8247
Acurácia do modelo_nb_pca: 0.3740
Acurácia do modelo_nb_lda: 0.9563
Acurácia do modelo_ad_normal: 0.9997
Acurácia do modelo_ad_pca: 0.7023
Acurácia do modelo_ad_lda: 0.9825
Acurácia do modelo_rf_normal: 0.9775
Acurácia do modelo_rf_pca: 0.7615
Acurácia do modelo_rf_lda: 0.9857


> Modelo de Árvore de Decisão treinado com dados originais (modelo_ad_normal) com acurácia de 99.97%.

> Modelo Random Forest treinado com dados originais (modelo_rf_normal) com acurácia de 97.75%.

> Modelo de Árvore de Decisão treinado com dados após LDA (modelo_ad_lda) com acurácia de 98.25%.